caikit/config/config.py

```diff
@@ -0,0 +1,154 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+"""Config methods for the `caikit` library. Mainly interacts with `config.yml`.
+"""
+
+# Standard
+from typing import Any, Dict, Optional
+import os
+import threading
+
+# First Party
+import aconfig
+import alog
+
+log = alog.use_channel("CONFIG")
+
+BASE_CONFIG_PATH = os.path.realpath(
+    os.path.join(os.path.dirname(__file__), "config.yml")
+)
+
+# The core config object that is continually merged into
+_CONFIG: aconfig.Config = aconfig.Config({})
+# An immutable view into the core config object, to be passed to callers
+_IMMUTABLE_CONFIG: aconfig.ImmutableConfig = aconfig.ImmutableConfig({})
+_CONFIG_LOCK: threading.Lock = threading.Lock()
+
+
+def get_config() -> aconfig.Config:
+    """Get the caikit configuration"""
+    return _IMMUTABLE_CONFIG
+
+
+def configure(
+    config_yml_path: Optional[str] = None, config_dict: Optional[Dict[str, Any]] = None
+):
+    """Configure caikit for your usage!
+    Merges into the internal aconfig.Config object with overrides from multiple sources.
+
+    Sources, last takes precedence:
+        1. The existing configuration from calls to `caikit.configure()`
+        2. The config from `config_yml_path`
+        3. The config from `config_dict`
+        4. The config files specified in the `config_files` configuration
+            (NB: This may be set by the `CONFIG_FILES` environment variable)
+        5. Environment variables, in ALL_CAPS_SNAKE_FORMAT
+
+    Args:
+        config_yml_path (Optional[str]): The path to the base configuration yaml
+            with overrides for your usage.
+        config_dict (Optional[Dict]): Config overrides in dictionary form
+
+    Returns: None
+        This only sets the config object that is returned by `caikit.get_config()`
+    """
+    if not config_yml_path and not config_dict:
+        log.error("<RUN43273054E>", "No config_file or config_dict provided")
+        raise ValueError("No config_file or config_dict provided")
+
+    cfg = aconfig.Config(_CONFIG)
+    if config_yml_path:
+        cfg = merge_configs(cfg, aconfig.Config.from_yaml(config_yml_path))
+    if config_dict:
+        cfg = merge_configs(cfg, aconfig.Config(config_dict))
+
+    cfg = _merge_extra_files(cfg)
+    _update_global_config(cfg)
+
+
+def _update_global_config(cfg: aconfig.Config):
+    """Updates the caikit config and creates a new immutable view of it to be shared via
+    get_config().
+    Locked because who the heck knows if merge_configs() is threadsafe.
+    """
+    # pylint: disable=global-statement
+    global _IMMUTABLE_CONFIG
+    # Update the config by merging the new updates over the existing config
+    with _CONFIG_LOCK:
+        # Locked just in case `configure()` is called concurrently for any reason
+        merge_configs(_CONFIG, cfg)
+        _IMMUTABLE_CONFIG = aconfig.ImmutableConfig(_CONFIG, override_env_vars=False)
```

Comment:
Very nice. Good catch on `override_env_vars=False` here. I can't imagine _how_ they would change, but it's nicely tight that way.